pkg/metric/writer_cw.go

```diff
@@ -31,101 +30,59 @@ const (
 )
 
 type (
-	StandardUnit = types.StandardUnit
-	Dimensions   map[string]string
+	cwWriterCtxKey string
+	Dimensions     map[string]string
+	StandardUnit   = types.StandardUnit
 )
 
-type Datum struct {
-	Priority   int          `json:"-"`
-	Timestamp  time.Time    `json:"timestamp"`
-	MetricName string       `json:"metricName"`
-	Dimensions Dimensions   `json:"dimensions"`
-	Value      float64      `json:"value"`
-	Unit       StandardUnit `json:"unit"`
-}
-
-func (d *Datum) Id() string {
-	return fmt.Sprintf("%s:%s", d.MetricName, d.DimensionKey())
-}
-
-func (d *Datum) DimensionKey() string {
-	dims := make([]string, 0)
-
-	for k, v := range d.Dimensions {
-		flat := fmt.Sprintf("%s:%s", k, v)
-		dims = append(dims, flat)
-	}
-
-	sort.Strings(dims)
-	dimKey := strings.Join(dims, "-")
-
-	return dimKey
-}
-
-func (d *Datum) IsValid() error {
-	if d.MetricName == "" {
-		return fmt.Errorf("missing metric name")
-	}
-
-	if d.Priority == 0 {
-		return fmt.Errorf("metric %s has no priority", d.MetricName)
-	}
-
-	if d.Unit == "" {
-		return fmt.Errorf("metric %s has no unit", d.MetricName)
-	}
-
-	return nil
-}
-
-type Data []*Datum
-
-type cwWriter struct {
+type cloudwatchWriter struct {
 	logger      log.Logger
 	clock       clock.Clock
 	client      gosoCloudwatch.Client
 	cwNamespace string
 }
 
-func NewCwWriter(ctx context.Context, config cfg.Config, logger log.Logger) (Writer, error) {
-	testClock := clock.NewRealClock()
-	settings := getMetricSettings(config)
-	cwNamespace := GetCloudWatchNamespace(config)
+func ProvideCloudwatchWriter(ctx context.Context, config cfg.Config, logger log.Logger) (Writer, error) {
```

Comment:
Usually, we have the 3 `Provide*`, `New* ` and `New*WithInterfaces`?\
Do we want to split it here as well again?